File: java-devildyw/src/main/java/com/dyw/leetcode/TOP面试题/L15三数之和.java
package com.dyw.leetcode.TOP面试题;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * @author Devil
 * @since 2022-12-09-17:40
 */
@SuppressWarnings("all")
public class L15三数之和 {
    public static void main(String[] args) {

    }

    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        Arrays.sort(nums);


        for (int i = 0; i < nums.length; i++) {
            if (nums[i] > 0) {
                //如果排序后的遍历到的元素大于0了 那么后续的元素都不会在小于0了 也就不会有可能等于0了
                return result;
            }
            if (i > 0 && nums[i - 1] == nums[i]) {
                continue;
            }

            int left = i + 1;
            int right = nums.length - 1;
            while (right > left) {
                int sum = nums[i] + nums[left] + nums[right];

                if (sum > 0) {
                    right--;
                } else if (sum < 0) {
                    left++;
                } else {
                    result.add(Arrays.asList(nums[i], nums[left], nums[right]));
                    //去重
                    while (right > left && nums[right] == nums[right - 1]) {
                        right--;
                    }
                    while (right > left && nums[left] == nums[left + 1]) {
                        left++;
                    }
                    right--;
                    left++;
                }
            }
        }
        return result;
    }
}

File: golang-devildyw/middle/L15三数之和.go
package main

import (
	"fmt"
	"sort"
)

func main() {

	for _, ints := range threeSum([]int{-1, 0, 1, 2, -1, -4}) {
		for _, num := range ints {
			fmt.Println(num)
		}
	}
}

func threeSum(nums []int) [][]int {
	// 先排序
	sort.Ints(nums)
	result := make([][]int, 0)

	for i := 0; i < len(nums); i++ {
		// 确保第一层不会重复了
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}

		l, r := i+1, len(nums)-1
		for l < r {
			sum := nums[i] + nums[l] + nums[r]
			if sum == 0 {
				result = append(result, []int{nums[i], nums[l], nums[r]})
				for l < r && nums[l] == nums[l+1] {
					l++
				}
				for l < r && nums[r] == nums[r-1] {
					r--
				}
				l++
				r--
			} else if sum < 0 {
				l++
			} else {
				r--
			}
		}
	}
	return result
}
